### PR TITLE
Yard doc'ing reek internals and improving documentation.

### PR DESCRIPTION
**Pull request by troessner**

### CODE REVIEW

**Comment by troessner** [comment]
I realize there is some duplication between "How reek works" and tree_dresser.rb.
I believe that's ok since both talk about the same thing but on a different level (the explanation in tree_dresser would be a source documentation whereas "How reek works" is an overview).

**Comment by troessner** [comment]
Pinging @chastell & @ mvz - anybody up for reviewing this?

File: lib/reek/tree_dresser.rb [hunk 1]
```diff
@@ -10,6 +10,47 @@ def initialize(klass_map = AST::ASTNodeClassMap.new)
...
+    #
+    # At the end when we're done, all nodes will have been recursively adorned by
+    # our SexpExtensions in a similar fashion.
+    # The root node class for instance would have the following, very similar
+    # ancestors chain:
+    # [ Reek::AST::SexpNode, Reek::AST::SexpExtensions::ClassNode,
+    #   Reek::AST::SexpExtensions::ModuleNode, Reek::AST::ASTNode, ... ]
+    # which resembles the one from [1]
+    # but instead of SexpExtensions::ConstNode we have SexpExtensions::ClassNode
+    # and SexpExtensions::ModuleNode mixed in
```
Comment (chastell) [inline lib/reek/tree_dresser.rb]: I’d add a period at the end of this sentence. :)

**Comment by chastell** [comment]
I’m on holidays this week, but I did a cursory check and made the worst nitpick ever. :)

I’ll merge this after that period is added, but wouldn’t mind @mvz reviewing it later anyway, as my `reek`-internals-fu is not super strong.

File: lib/reek/tree_dresser.rb [hunk 2]
```diff
@@ -10,6 +10,47 @@ def initialize(klass_map = AST::ASTNodeClassMap.new)
       @klass_map = klass_map
     end
 
+    # Given this sexp (Parser::AST::Node):
```
Comment (mvz) [inline lib/reek/tree_dresser.rb]: For this comment, I'm missing the big picture a bit here. How about starting with something like:

> Recursively enhance an AST with type-dependent mixins, and comments.

File: lib/reek/tree_dresser.rb [hunk 3]
```diff
@@ -10,6 +10,47 @@ def initialize(klass_map = AST::ASTNodeClassMap.new)
...
+    #
+    # Let's just look at what happens with [1]:
+    #   - `type` is going to be `const`
+    #   - children would be [nil, :C]
+    #   - the resulting class from @klass_map.klass_for(type)
+    #     would be an anonymous class with Reek::AST::ASTNode as superclass,
+    #     so not Parser::AST::Node anymore and it would have included
+    #     Reek::AST::SexpExtensions::ConstNode.
+    #     The overall ancestors chain just for [1] would look like this:
+    #     [ Reek::AST::SexpNode, Reek::AST::SexpExtensions::ConstNode, Reek::AST::ASTNode,
```
Comment (mvz) [inline lib/reek/tree_dresser.rb]: I'm a little worried that all these details can become out-of-sync with reality very quickly. Can we explain all this in a more general way?

File: lib/reek/source/source_code.rb [hunk 4]
```diff
@@ -53,7 +81,15 @@ def syntax_tree
...
             end
 
+            # See https://whitequark.github.io/parser/Parser/Source/Comment/Associator.html
             comment_map = Parser::Source::Comment.associate(ast, comments) if ast
+            # The AST (Parser::AST::Node) at this stage would look like the one in the method
+            # documentation above (but without the SexpExtensions which will be added next).
+            # A comment_map would look like
+            # { (class (const nil :C).... =>
+            #     [#<Parser::Source::Comment string:1:9 "# comment about C">]}
+            # With the key being the AST and the value an array with all relevant comments.
```
Comment (mvz) [inline lib/reek/source/source_code.rb]: Won't this be clear from the page linked to above?

File: lib/reek/source/source_code.rb [hunk 5]
```diff
@@ -41,9 +41,37 @@ def self.from(source)
...
+      #
+      #   (class
+      #     (const nil :C) nil
+      #     (def :m
+      #       (args)
+      #       (send nil :puts
+      #         (str "nada"))))
+      #
+      # where each node is possibly adorned with our SexpExtensions (see ast/ast_node_class_map
+      # and ast/sexp_extensions for details).
```
Comment (mvz) [inline lib/reek/source/source_code.rb]: This is very clear :+1:

File: lib/reek/tree_dresser.rb [hunk 6]
```diff
@@ -10,6 +10,47 @@ def initialize(klass_map = AST::ASTNodeClassMap.new)
...
+    # [ Reek::AST::SexpNode, Reek::AST::SexpExtensions::ClassNode,
+    #   Reek::AST::SexpExtensions::ModuleNode, Reek::AST::ASTNode, ... ]
+    # which resembles the one from [1]
+    # but instead of SexpExtensions::ConstNode we have SexpExtensions::ClassNode
+    # and SexpExtensions::ModuleNode mixed in
+
+    # @param sexp [Parser::AST::Node] - the given sexp
+    # @param comment_map [Hash] - see the documentation for SourceCode#syntax_tree
+    #
+    # @return an instance of Reek::AST::SexpNode with type-dependent sexp extensions mixed in.
```
Comment (mvz) [inline lib/reek/tree_dresser.rb]: I *think* this should be `Reek::AST::ASTNode`. Boy, this is confusing. Good thing we're already working toward simplification.

File: docs/How-reek-works-internally.md [hunk 7]
```diff
@@ -54,3 +56,60 @@
...
+          nil
+         (def :m              [AST::SexpExtensions::DefNode, AST::SexpExtensions::MethodNodeBase]
+           (args)             [AST::SexpExtensions::ArgsNode]
+           (send nil :puts    [AST::SexpExtensions::SendNode]
+             (str "nada"))))
+            |
+            |
+            |
+      A TreeWalker then traverses this now adorned tree again and
+      runs all SmellDetectors from the SmellRepository above
```
Comment (mvz) [inline docs/How-reek-works-internally.md]: I haven't checked how this looks once YARD is done with it, but I think this is very clear. Perhaps we can just point here from the method comment of `TreeDresser#dress`?

File: lib/reek/tree_dresser.rb (hunk 6 above)
Comment (troessner) [inline lib/reek/tree_dresser.rb]: >> I think this should be Reek::AST::ASTNode. 

Good catch! Indeed, it should be, changed.:)

>> Boy, this is confusing. Good thing we're already working toward simplification.

I started a discussion regarding this here: https://github.com/troessner/reek/issues/542

File: lib/reek/tree_dresser.rb (hunk 1 above)
Comment (troessner) [inline lib/reek/tree_dresser.rb]: Fixed.:)

File: lib/reek/tree_dresser.rb (hunk 3 above)
Comment (troessner) [inline lib/reek/tree_dresser.rb]: Good point. I'll just remove this whole section except for your comment above and yard-doc and point to: docs/How-reek-works-internally.md

File: docs/How-reek-works-internally.md (hunk 7 above)
Comment (troessner) [inline docs/How-reek-works-internally.md]: Done.

File: lib/reek/source/source_code.rb (hunk 4 above)
Comment (troessner) [inline lib/reek/source/source_code.rb]: Yes, you are right.

**Comment by troessner** [comment]
Adressed all your comments @mvz && @chastell !

**Pull request merged**